llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -1516,16 +1516,23 @@ static void transformRecipestoEVLRecipes(VPlan &Plan, VPValue &EVL) {
                         IntegerType::getInt1Ty(CI->getContext())));
                     Ops.push_back(Mask);
                     Ops.push_back(&EVL);
+                    FastMathFlags FMF = {};
+                    if (isa<FPMathOperator>(CI))
+                      FMF = CI->getFastMathFlags();
                     return new VPWidenIntrinsicRecipe(
-                        VPID, Ops, TypeInfo.inferScalarType(CInst),
+                        VPID, Ops, TypeInfo.inferScalarType(CInst), FMF,
                         CInst->getDebugLoc());
                   })
               .Case<VPWidenSelectRecipe>([&](VPWidenSelectRecipe *Sel) {
+                auto *SelInst = dyn_cast<SelectInst>(Sel->getUnderlyingInstr());
```

Comment:
Please only use flags from the recipe, not the underlying instruction. Using from the underlying instruction may be incorrect in some cases, e.g. if the recipe originally was only execute unconditionally.

Same for `VPWidenCastRecipe` above.


Maybe start with just `Select` VPInstructions here, then PRs to add flags to `VPWidenCastRecipe`/`VPWidenSelectRecipe` separately, then a PR to pass them through to the intrinsics.

Reply:
Thank you very much for your patience. Wait until #119510 is processed properly, and I will correct it.

Reply:
Sorry, I don't quite understand. ”Maybe start with just Select VPInstructions here“ ， VPWidenSelectRecipe inherit VPSingleDefRecipe， It seems that the only way to get the flags is by underlying instruction. Or  change VPWidenSelectRecipe inheritance to VPRecipeWithIRFlags, do you think it is ok?

Reply:
> Or change VPWidenSelectRecipe inheritance to VPRecipeWithIRFlags, do you think it is ok?

+1

Reply:
pls review： #121023